mobile/library/jni/jni_impl.cc

```diff
@@ -828,46 +763,113 @@ extern "C" JNIEXPORT jint JNICALL Java_io_envoyproxy_envoymobile_engine_JniLibra
   auto java_stream_callbacks_global_ref = jni_helper.newGlobalRef(java_stream_callbacks).release();
   auto engine = reinterpret_cast<Envoy::InternalEngine*>(engine_handle);
   Envoy::EnvoyStreamCallbacks stream_callbacks;
-  // TODO(fredyw): Rewrite the whole implementation to not use the bridge data structures.
-  stream_callbacks.on_headers_ =
-      [java_stream_callbacks_global_ref](const Envoy::Http::ResponseHeaderMap& headers,
-                                         bool end_stream, envoy_stream_intel stream_intel) {
-        envoy_headers bridge_headers = Envoy::Http::Utility::toBridgeHeaders(headers);
-        jvm_on_response_headers(bridge_headers, end_stream, stream_intel,
-                                java_stream_callbacks_global_ref);
-      };
+  stream_callbacks.on_headers_ = [java_stream_callbacks_global_ref](
+                                     const Envoy::Http::ResponseHeaderMap& headers, bool end_stream,
+                                     envoy_stream_intel stream_intel) {
+    Envoy::JNI::JniHelper jni_helper(Envoy::JNI::JniHelper::getThreadLocalEnv());
+    auto java_headers = Envoy::JNI::cppHeadersToJavaHeaders(jni_helper, headers);
+    auto java_stream_intel = Envoy::JNI::cppStreamIntelToJavaStreamIntel(jni_helper, stream_intel);
+    auto java_stream_callbacks_class = jni_helper.getObjectClass(java_stream_callbacks_global_ref);
+    auto java_on_headers_method_id = jni_helper.getMethodId(
+        java_stream_callbacks_class.get(), "onHeaders",
+        "(Ljava/util/Map;ZLio/envoyproxy/envoymobile/engine/types/EnvoyStreamIntel;)V");
```

Comment:
I see that we're looking a lot of methods up now by their exact argument types, whereas before it seems we'd look up the same methods without that... is this something required by the new jclass cache?  I think I'm not following the new logic here, but I knew very little about JNI so I'm mostly asking for educational purposes.

Reply:
If you look at the old code, basically we are passing `byte[] key, byte[] value` (https://github.com/envoyproxy/envoy/blob/266f9a35ab9d698ec807999c7d5bfd226bdeb8bf/mobile/library/java/io/envoyproxy/envoymobile/engine/JvmCallbackContext.java#L25-L27) and converting them into a `Map` (https://github.com/envoyproxy/envoy/blob/266f9a35ab9d698ec807999c7d5bfd226bdeb8bf/mobile/library/java/io/envoyproxy/envoymobile/engine/JvmBridgeUtility.java#L57-L62).

The old code uses intermediate types and has more translation layers, which this PR is trying to eliminate. For the `onHeader` case, it will first create a pair of `byte[]`s and then convert them into Java `Map`, which I personally think is wasteful and not to mention the code isn't very easy to read.